Move to Kubernetes 1.11

dlipovetsky opened the pull request.

Commit:
nodeadm: Prevent nil dereference of VIPConfiguration or ClusterConfiguration

Commit:
nodeadm: Configure NodeRegistration and factor our network validation

Commit:
nodeadm: Remove deprecated apiserver --allow-privileged flag

Commit:
Move kubeadm types to separate file

Commit:
Generate metadata for kubeadm config

Commit:
Use configuration file for kubeadm discovery and bootstrap configuration

Commit:
Make nodeadm configuration readable by user only

The join configuration contains a token that should be kept secret.
The master configuration contains configuration options that may need to be
kept secret.

Commit:
Set kubernetesVersion in kubeadm config

Commit:
Update etcd configuration

ojmhetar commented on pkg/nodeadm/nodeadm.go:
```diff
@@ -164,44 +121,48 @@ func setInitConfigFromClusterConfig(cfg *InitConfiguration, clusterConfig *spv1.
+
+func validateClusterNetworkingConfiguration(cluster clusterv1.Cluster) error {
+	switch cbl := len(cluster.Spec.ClusterNetwork.Pods.CIDRBlocks); {
+	case cbl < 1:
```
why not just cbl != 1 here?

dlipovetsky replied:
I think the error is more helpful when it is specific ("at most one" or "at least one").

If I want the more helpful error and I use `cbl != 1` here, I'll still have to handle the two cases separately.

Commit:
Fix kubeadm schema

KubeletConfiguration and KubeProxyConfiguration were missing the
parent fields under which they are nested in kubeadm v1alpha2 types.

Commit:
Omit a nil VIPConfiguration from the nodeadm configuration

Commit:
Implement set for Cluster.Spec.APIEndpoints

puneetguptanitj commented on pkg/nodeadm/kubeadm.go:
```diff
@@ -0,0 +1,114 @@
+	Etcd                       Etcd                 `json:"etcd,omitempty"`
+	KubernetesVersion          string               `json:"kubernetesVersion,omitempty"`
+	Networking                 Networking           `json:"networking,omitempty"`
+	KubeletConfiguration       KubeletConfiguration `json:"kubeletConfiguration,omitempty"`
```
Hi Daniel, just to understand, you'd mentioned that using upstream types adds noise in the serialized yaml. would that be an issue here as well.

dlipovetsky replied:
I agree. For this, I filed https://github.com/platform9/ssh-provider/issues/61 some time back.

Because ssh-provider does not introspect KubeletConfiguration or KubeProxyConfiguration, it might be reasonable to make both `map[string]interface{}` until the types are fixed upstream. I'll follow up with the above issue.

puneetguptanitj commented on pkg/nodeadm/nodeadm.go:
```diff
@@ -35,89 +34,61 @@ type VIPConfiguration struct {
-		cfg.Networking.ServiceSubnet = cluster.Spec.ClusterNetwork.Services.CIDRBlocks[0]
-	case 2:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at most one block", cluster.Name)
+	if err := validateClusterNetworkingConfiguration(cluster); err != nil {
```
Since these would be passed through MasterConfiguration and processed by kubeadm, should nodeadm be doing this verification, or can it be offloaded to kubeadm as well?

dlipovetsky replied:
This validates Cluster configuration, which ssh provider must transform to kubeadm configuration.

This is exactly [what the AWS provider does](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/870d03d2d5a25a407b335d3f0b702227dfd7cd59/pkg/cloud/aws/services/kubeadm/aws_defaults.go#L85), though it doesn't validate the Cluster configuration, yet.

puneetguptanitj reviewed (commented):
Hi Daniel, lgtm, had some questions

dlipovetsky commented on pkg/util/sets/apiendpoint.go:
```diff
@@ -0,0 +1,73 @@
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+type APIEndpointSet map[clusterv1.APIEndpoint]struct{}
```
The SSH provider controller, once available, would update Cluster.Status.APIEndpoints.
